src/server.rs

```diff
@@ -0,0 +1,221 @@
+            "ID" => Ok(ClientVersion),
+            "CH" => Ok(KeepAlive),
+            "askchaa" => Ok(AskListLengths),
+            "askchar" => Ok(AskListCharacters),
+            "AN" => Ok(CharacterList),
+            "AE" => Ok(EvidenceList),
+            "AM" => Ok(MusicList),
+            "RC" => Ok(AO2CharacterList),
+            "RD" => Ok(AO2Ready),
+            "CC" => Ok(SelectCharacter),
+            "MS" => Ok(ICMessage),
+            "CT" => Ok(OOCMessage),
+            "MC" => Ok(PlaySong),
+            "RT" => Ok(WTCEButtons),
+            "SETCASE" => Ok(SetCasePreferences),
+            "CASEA" => Ok(CaseAnnounce),
+            "HP" => Ok(Penalties),
+            "PE" => Ok(AddEvidence),
+            "DE" => Ok(DeleteEvidence),
+            "EE" => Ok(EditEvidence),
+            "ZZ" => Ok(CallModButton),
+            "opKICK" => Ok(KickWithGuard),
+            "opBAN" => Ok(BanWithGuard),
+            _ => anyhow::bail!("Invalid command!"),
+        }
+    }
+}
+
+enum DecodeState {
+    Command,
+    Arguments,
+}
+
+use serde::export::Formatter;
+use std::{
+    io::{BufRead, Cursor, Seek, SeekFrom},
+    str::FromStr,
+};
+use tokio::{
+    io::{AsyncBufReadExt, Error, ErrorKind},
+    stream::StreamExt,
+};
+use tokio_util::codec::{Decoder, Encoder, FramedRead};
+
+#[derive(Debug)]
+pub struct AOMessage {
+    pub command: Command,
+    pub args: Vec<String>,
+}
+
+pub struct AOMessageCodec {
+    state: DecodeState,
+}
+
+impl AOMessageCodec {
+    pub fn new() -> Self {
+        Self { state: DecodeState::Command }
+    }
+}
+
+impl Decoder for AOMessageCodec {
+    type Item = AOMessage;
+    type Error = anyhow::Error;
+
+    fn decode(
+        &mut self,
+        mut src: &mut BytesMut,
+    ) -> Result<Option<Self::Item>, Self::Error> {
+        let mut reader = Cursor::new(src);
+        let mut cmd_buf = Vec::with_capacity(8);
+        let cmd_len = BufRead::read_until(&mut reader, b'#', &mut cmd_buf)?;
+
+        if cmd_len == 0 {
+            log::debug!("Invalid protocol?!");
+            anyhow::bail!("Invalid protocol!")
+        }
+
+        cmd_buf.truncate(cmd_len.saturating_sub(1));
+        let command: Command = unsafe {
+            // We are sure that the commands will *always* be valid utf8
+            String::from_utf8_unchecked(cmd_buf)
+        }
+        .parse()?;
+
+        reader.seek(SeekFrom::Start(cmd_len as u64))?;
+        let mut splitted = BufRead::split(reader, b'#')
+            .map(|l| String::from_utf8(l.unwrap()).unwrap());
+        let mut args: Vec<String> = splitted.collect();
+        let mut end_header_index = 0;
+
+        for (i, arg) in args.iter().enumerate() {
+            if arg.chars().nth(0).unwrap() == '%' {
+                end_header_index = i;
+                break;
+            } else {
+                continue;
+            }
+        }
+
+        args.remove(end_header_index);
+
+        Ok(Some(AOMessage { command, args: vec![] }))
```

Comment:
Didn't you forget to remove `vec![]`?
```suggestion
        Ok(Some(AOMessage { command, args }))
```